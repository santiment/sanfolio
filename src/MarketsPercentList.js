import React from 'react'
import { connect } from 'react-redux'
import {
  Button
} from 'semantic-ui-react'
import {
  pure,
  compose,
  lifecycle,
  withState
} from 'recompose'
import { formatNumber } from './utils/formatting'

const getPercent = mPercent => mPercent[Object.keys(mPercent)[0]]

const getSymbol = mPercent => Object.keys(mPercent)[0]

export const searchPriceCoinBySymbol = (symbol, prices) => {
  const price = prices.find(price => {
    return Object.keys(price)[0] === symbol
  })
  if (!price) {
    // TODO: we don't have all prices for top 20 market cap!
    // return 1 for this. IT's critical place!
    return 1
  }
  return Object.values(price)[0]
}

export const AssetRow = ({
  cap,
  money,
  prices
}) => {
  const symbol = getSymbol(cap)
  const coinMoney = getPercent(cap) / 100 * money
  const coin = coinMoney / searchPriceCoinBySymbol(symbol, prices)
  return (
    <div className='markets-percent-list-row'>
      <div className='markets-percent-list-row-symbol'>
        {getSymbol(cap)} - {getPercent(cap)}%
      </div>
      <div className='markets-percent-list-coin-value'>
        {coin.toFixed(8)}
      </div>
      <div className='markets-percent-list-row-value'>
        {money > 0 && formatNumber((coinMoney).toFixed(2), 'USD')}
      </div>
    </div>
  )
}

export const MarketsPercentList = ({
  markets,
  money,
  prices,
  onListApproved,
  onListDeclined,
  data
}) => {
  return (
    <div className='markets-percent-list'>
      <div className='markets-percent-list-inner'>
        {markets.map((cap, index) => (
          <AssetRow
            key={index}
            money={money}
            prices={prices}
            cap={cap} />
        ))}
      </div>
      <div className='markets-percent-list-control'>
        <Button
          onClick={onListDeclined}
          basic
          color='red'>
          Reset
        </Button>
        <Button
          onClick={() => onListApproved(data)}
          color='green'>
          Confirm balanced investing schema
        </Button>
      </div>
    </div>
  )
}

const mapStateToProps = state => {
  const prices = Object.keys(state.prices.items).map(key => {
    return {
      [key]: state.prices.items[key][0]
    }
  })
  return {
    prices
  }
}

const enhance = compose(
  pure,
  withState('data', '_', {}),
  connect(
    mapStateToProps
  ),
  lifecycle({
    componentDidMount () {
      const {
        markets,
        money,
        prices
      } = this.props
      const data = markets.map((cap, index) => {
        const symbol = getSymbol(cap)
        const coinMoney = getPercent(cap) / 100 * money
        const coin = coinMoney / searchPriceCoinBySymbol(symbol, prices)
        return {
          [symbol]: coin
        }
      }).reduce((acc, val, index) => {
        acc[Object.keys(val)[0]] = Object.values(val)[0]
        return acc
      }, {})
      this.setState({data})
    }
  })
)

export default enhance(MarketsPercentList)
