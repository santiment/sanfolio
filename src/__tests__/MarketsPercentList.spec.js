/* eslint-env jasmine */
import { MarketsPercentList, AssetRow } from '../MarketsPercentList.js'
import React from 'react'
import { configure, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
configure({ adapter: new Adapter() })

describe('MarketsPercentList', () => {
  const markets = [
    {'BTC': 56.60},
    {'ETH': 20.00},
    {'ETC': 12.12}
  ]

  const data = {
    ADA: 11.609008590666358,
    BCH: 0.007558669061380139,
    BTC: 0.00750053228379006,
    DASH: 0.0034625156131839,
    EOS: 0.2208225112777211,
    ETC: 0.04403520558230914,
    ETH: 0.04307733540298107,
    HSR: 0.01925753932664638,
    LSK: 0.051347106439531234,
    LTC: 0.024265747042433658,
    MIOTA: 1.2532610657778096,
    NEO: 0.02912821789038175
  }

  const money = 250

  const prices = [
    {'BTC': 7700},
    {'ETH': 330},
    {'ETC': 66}
  ]

  it('should render correctly', () => {
    const wrapper = mount(
      <MarketsPercentList
        markets={markets}
        prices={prices}
        data={data}
        money={money} />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  describe('AssetRow', () => {
    it('should render correctly', () => {
      const wrapper = mount(
        <AssetRow
          index={0}
          cap={{
            'BTC': 56.60
          }}
          data={data}
          prices={prices}
          money={money} />
      )
      expect(toJson(wrapper)).toMatchSnapshot()
    })
  })
})
