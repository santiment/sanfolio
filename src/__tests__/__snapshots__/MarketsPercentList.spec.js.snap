// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketsPercentList AssetRow should render correctly 1`] = `
<Component
  cap={
    Object {
      "BTC": 56.6,
    }
  }
  data={
    Object {
      "ADA": 11.609008590666358,
      "BCH": 0.007558669061380139,
      "BTC": 0.00750053228379006,
      "DASH": 0.0034625156131839,
      "EOS": 0.2208225112777211,
      "ETC": 0.04403520558230914,
      "ETH": 0.04307733540298107,
      "HSR": 0.01925753932664638,
      "LSK": 0.051347106439531234,
      "LTC": 0.024265747042433658,
      "MIOTA": 1.2532610657778096,
      "NEO": 0.02912821789038175,
    }
  }
  index={0}
  money={250}
  prices={
    Array [
      Object {
        "BTC": 7700,
      },
      Object {
        "ETH": 330,
      },
      Object {
        "ETC": 66,
      },
    ]
  }
>
  <div
    className="markets-percent-list-row"
  >
    <div
      className="markets-percent-list-row-symbol"
    >
      BTC
       - 
      56.6
      %
    </div>
    <div
      className="markets-percent-list-coin-value"
    >
      0.00750053
    </div>
    <div
      className="markets-percent-list-row-value"
    >
      $0.00
    </div>
  </div>
</Component>
`;

exports[`MarketsPercentList should render correctly 1`] = `
<Component
  data={
    Object {
      "ADA": 11.609008590666358,
      "BCH": 0.007558669061380139,
      "BTC": 0.00750053228379006,
      "DASH": 0.0034625156131839,
      "EOS": 0.2208225112777211,
      "ETC": 0.04403520558230914,
      "ETH": 0.04307733540298107,
      "HSR": 0.01925753932664638,
      "LSK": 0.051347106439531234,
      "LTC": 0.024265747042433658,
      "MIOTA": 1.2532610657778096,
      "NEO": 0.02912821789038175,
    }
  }
  markets={
    Array [
      Object {
        "BTC": 56.6,
      },
      Object {
        "ETH": 20,
      },
      Object {
        "ETC": 12.12,
      },
    ]
  }
  money={250}
  prices={
    Array [
      Object {
        "BTC": 7700,
      },
      Object {
        "ETH": 330,
      },
      Object {
        "ETC": 66,
      },
    ]
  }
>
  <div
    className="markets-percent-list"
  >
    <div
      className="markets-percent-list-inner"
    >
      <Component
        cap={
          Object {
            "BTC": 56.6,
          }
        }
        data={
          Object {
            "ADA": 11.609008590666358,
            "BCH": 0.007558669061380139,
            "BTC": 0.00750053228379006,
            "DASH": 0.0034625156131839,
            "EOS": 0.2208225112777211,
            "ETC": 0.04403520558230914,
            "ETH": 0.04307733540298107,
            "HSR": 0.01925753932664638,
            "LSK": 0.051347106439531234,
            "LTC": 0.024265747042433658,
            "MIOTA": 1.2532610657778096,
            "NEO": 0.02912821789038175,
          }
        }
        key="0"
        prices={
          Array [
            Object {
              "BTC": 7700,
            },
            Object {
              "ETH": 330,
            },
            Object {
              "ETC": 66,
            },
          ]
        }
      >
        <div
          className="markets-percent-list-row"
        >
          <div
            className="markets-percent-list-row-symbol"
          >
            BTC
             - 
            56.6
            %
          </div>
          <div
            className="markets-percent-list-coin-value"
          >
            0.00750053
          </div>
          <div
            className="markets-percent-list-row-value"
          >
            $0.00
          </div>
        </div>
      </Component>
      <Component
        cap={
          Object {
            "ETH": 20,
          }
        }
        data={
          Object {
            "ADA": 11.609008590666358,
            "BCH": 0.007558669061380139,
            "BTC": 0.00750053228379006,
            "DASH": 0.0034625156131839,
            "EOS": 0.2208225112777211,
            "ETC": 0.04403520558230914,
            "ETH": 0.04307733540298107,
            "HSR": 0.01925753932664638,
            "LSK": 0.051347106439531234,
            "LTC": 0.024265747042433658,
            "MIOTA": 1.2532610657778096,
            "NEO": 0.02912821789038175,
          }
        }
        key="1"
        prices={
          Array [
            Object {
              "BTC": 7700,
            },
            Object {
              "ETH": 330,
            },
            Object {
              "ETC": 66,
            },
          ]
        }
      >
        <div
          className="markets-percent-list-row"
        >
          <div
            className="markets-percent-list-row-symbol"
          >
            ETH
             - 
            20
            %
          </div>
          <div
            className="markets-percent-list-coin-value"
          >
            0.04307734
          </div>
          <div
            className="markets-percent-list-row-value"
          >
            $0.00
          </div>
        </div>
      </Component>
      <Component
        cap={
          Object {
            "ETC": 12.12,
          }
        }
        data={
          Object {
            "ADA": 11.609008590666358,
            "BCH": 0.007558669061380139,
            "BTC": 0.00750053228379006,
            "DASH": 0.0034625156131839,
            "EOS": 0.2208225112777211,
            "ETC": 0.04403520558230914,
            "ETH": 0.04307733540298107,
            "HSR": 0.01925753932664638,
            "LSK": 0.051347106439531234,
            "LTC": 0.024265747042433658,
            "MIOTA": 1.2532610657778096,
            "NEO": 0.02912821789038175,
          }
        }
        key="2"
        prices={
          Array [
            Object {
              "BTC": 7700,
            },
            Object {
              "ETH": 330,
            },
            Object {
              "ETC": 66,
            },
          ]
        }
      >
        <div
          className="markets-percent-list-row"
        >
          <div
            className="markets-percent-list-row-symbol"
          >
            ETC
             - 
            12.12
            %
          </div>
          <div
            className="markets-percent-list-coin-value"
          >
            0.04403521
          </div>
          <div
            className="markets-percent-list-row-value"
          >
            $0.00
          </div>
        </div>
      </Component>
    </div>
    <div
      className="markets-percent-list-control"
    >
      <Button
        as="button"
        basic={true}
        color="red"
      >
        <button
          className="ui red basic button"
          onClick={[Function]}
          role="button"
        >
          Reset
        </button>
      </Button>
      <Button
        as="button"
        color="green"
        onClick={[Function]}
      >
        <button
          className="ui green button"
          onClick={[Function]}
          role="button"
        >
          Confirm balanced investing schema
        </button>
      </Button>
    </div>
  </div>
</Component>
`;
